python/lsst/cp/pipe/bfCoeffCalc.py

```diff
@@ -0,0 +1,1272 @@
+
+        Parameters
+        ----------
+        dataRef : list of lsst.daf.persistence.ButlerDataRef
+            dataRef for the CCD for the visits to be fit.
+        visitPairs : `iterable` of `tuple` of `int`
+            Pairs of visit numbers to be processed together
+        """
+        xcorrs = {}  # dict of lists keyed by either amp or CCD depending on level
+        means = {}
+        kernels = {}
+
+        # setup necessary objects
+        ccdNum = dataRef.dataId[self.config.ccdKey]
+        if self.config.level == 'CCD':
+            xcorrs = {ccdNum: []}
+            means = {ccdNum: []}
+        elif self.config.level == 'AMP':
+            # NB: don't use dataRef.get('raw_detector') due to composites
+            detector = dataRef.get('camera')[dataRef.dataId[self.config.ccdKey]]
+            ampInfoCat = detector.getAmpInfoCatalog()
+            ampNames = [amp.getName() for amp in ampInfoCat]
+            xcorrs = {key: [] for key in ampNames}
+            means = {key: [] for key in ampNames}
+
+        # calculate or retrieve the gains
+        if self.config.doCalcGains:
+            self.log.info('Beginning gain estimation for CCD %s'%ccdNum)
+            gains, nomGains = self.estimateGains(dataRef, visitPairs)
+            dataRef.put(gains, datasetType='bfGain')
+            self.log.info('Finished gain estimation for CCD %s'%ccdNum)
+        else:
+            gains = dataRef.get('bfGain')
+            if not gains:
+                self.log.fatal('Failed to retrieved gains for CCD %s'%ccdNum)
+                raise RuntimeError("Must either calculate or supply gains for %s"%ccdNum)
+            self.log.info('Retrieved stored gain for CCD %s'%ccdNum)
+            self.log.debug('Gains for CCD %s: %s'%(ccdNum, gains))
+        self.log.debug('CCD %s has gains %s'%(ccdNum, gains))
+
+        # Loop over pairs of visits, calculating the cross correlations at the required level
+        for (v1, v2) in visitPairs:
+
+            dataRef.dataId['visit'] = v1
+            exp1 = self.isr.runDataRef(dataRef).exposure
+            dataRef.dataId['visit'] = v2
+            exp2 = self.isr.runDataRef(dataRef).exposure
+            del dataRef.dataId['visit']
+
+            self.log.info('Preparring images for cross correlation calculation for CCD %s'%ccdNum)
+            # note the shape of these returns depends on level
+            _scaledMaskedIm1, _means1 = self._prepareImage(exp1, gains, self.config.level)
+            _scaledMaskedIm2, _means2 = self._prepareImage(exp2, gains, self.config.level)
+
+            if self.debug.enabled:
+                try:
+                    frameId1 = exp1.getMetadata().get("FRAMEID")
+                    frameId2 = exp2.getMetadata().get("FRAMEID")
+                    frameId = '_diff_'.join(frameId1, frameId2)
+                except:
+                    frameId = 'Im1 diff Im2'
+
+            # depending on level this is either one pass or n_amps
```

Comment:
This is cryptic and duplicates the comment on the next line. and, to some extent, the comment at the top of the loop
I suggest just one comment that is clearer, or one here and one at the top of the loop. I think you are trying to say something like this (though I may have it wrong and you certainly should reword it as you see fit):

    # Compute cross-correlation and means at the appropriate config.level:
    # - "DETECTOR": compare the two visits
    # - "AMP": compare each amplifier of one visit to ... (what? each amplifier of the other visit? the same amplifier of the other visit?)

Reply:
Yes, this was just a garbage comment, thanks. It not reads:

```
            # Compute the cross-correlation and means at the appropriate config.level:
            # - "DETECTOR": one key, so compare the two visits to each other
            # - "AMP": n_amp keys, comparing each amplifier of one visit to the same amplifier in other visit```